apis/cloudflare/src/attachment-wrapper.ts

```diff
@@ -0,0 +1,165 @@
+}
+
+function replacePayloadWithAttachmentsInner(
+  data: unknown,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): unknown {
+  if (isArray(data)) {
+    return data.map((item) =>
+      replacePayloadWithAttachmentsInner(item, state, attachments),
+    );
+  } else if (isObject(data)) {
+    return Object.fromEntries(
+      Object.entries(data).map(([key, value]) => [
+        key,
+        replacePayloadWithAttachmentsInner(value, state, attachments),
+      ]),
+    );
+  } else if (typeof data === "string") {
+    if (attachments[data]) {
+      return attachments[data];
+    } else if (
+      isBase64File(
+        data,
+        (mimeType) =>
+          mimeType.startsWith("image/") || mimeType === "application/pdf",
+      )
+    ) {
+      const { mimeType, data: arrayBuffer } = getBase64Parts(data);
+      const filename = `file.${mimeType.split("/")[1]}`;
+      return new Attachment({
+        data: arrayBuffer,
+        contentType: mimeType,
+        filename,
+        state,
+      });
+    } else {
+      return data;
+    }
+  } else {
+    return data;
+  }
+}
+
+export function replaceAttachmentsInPayload<T>(
+  data: T,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): Promise<T> {
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return replaceAttachmentsInPayloadInner(
+    data,
+    state,
+    attachments,
+  ) as Promise<T>;
+}
+
+async function replaceAttachmentsInPayloadInner(
+  payload: unknown,
+  state: BraintrustState | undefined,
+  attachments: Record<string, AttachmentReference>,
+): Promise<unknown> {
+  if (isArray(payload)) {
+    return Promise.all(
+      payload.map((item) =>
+        replaceAttachmentsInPayloadInner(item, state, attachments),
+      ),
+    );
+  } else if (payload instanceof ReadonlyAttachment) {
+    const ret = await replaceAttachmentReferenceObjectWithDownloadUrl(payload);
+    if (ret) {
+      attachments[ret] = payload.reference;
+    }
+    return ret;
+  } else if (isObject(payload)) {
+    const replaced = await replaceAttachmentReferenceWithDownloadUrl(
+      payload,
+      state,
+      attachments,
+    );
+    if (replaced) {
+      return replaced;
+    }
+
+    return Object.fromEntries(
+      await Promise.all(
+        Object.entries(payload).map(async ([key, value]) => [
+          key,
+          await replaceAttachmentsInPayloadInner(value, state, attachments),
+        ]),
+      ),
+    );
+  } else if (typeof payload === "string") {
+    const deserialized = deserializePlainStringAsJSON(payload).value;
+    return (
+      (await replaceAttachmentReferenceWithDownloadUrl(
+        deserialized,
+        state,
+        attachments,
+      )) ?? payload
+    );
+  }
+
+  return payload;
+}
+
+async function replaceAttachmentReferenceWithDownloadUrl(
+  payload: unknown,
```

Comment:
should this be the z.infer< of the braintrustAttachmentReferenceSchema?

Reply:
this is the routine that determines if the unknown type payload _is_ an attachment reference